### PR TITLE
Add Points collection observability support for Polygon and Polyline

### LINKED ISSUE #3623: Polygon and Polyline Points binding problem
Avalonia 0.9.3

Xaml:
```
<Polygon Fill="Red" Points="{Binding DeltaPoints}"/>
<Polyline Stroke="Blue" Points="{Binding DeltaPoints}"/>
```

C#:
```
private readonly ReadOnlyObservableCollection<Avalonia.Point> _deltaPoints;
public ReadOnlyObservableCollection<Avalonia.Point> DeltaPoints => _deltaPoints;
private SourceList<Avalonia.Point> _pointsD;

_pointsD = new SourceList<Avalonia.Point>();
            _pointsD.Add(new Avalonia.Point(123, 45));
            _pointsD.Add(new Avalonia.Point(45, 45));
           ...
            _pointsD.Connect()
                .ObserveOn(RxApp.MainThreadScheduler)
                .Bind(out _deltaPoints)
                .Subscribe();
```
These bindings work only for the initial points collection. All dynamically added points change neither Polygon nor Polyline image. Binding of the same collection to a ListBox works well.

### PR DESCRIPTION
**Pull request by SKProCH**
## What does the pull request do?
Add Points binding observability support for Polygon and Polyline

## What is the current behavior?
When ObservableCollection was bound to Points property, updating collection does not updates the geometry. See https://github.com/AvaloniaUI/Avalonia/issues/3623

## What is the updated/expected behavior with this PR?
When ObservableCollection was bound to Points property, updating collection updates the geometry (and control actially re-rendered with a new geometry)

## Checklist

- [x] Added unit tests (if possible)? (not possible)
- [x] Added XML documentation to any related classes?
- [x] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
None

## Fixed issues
Fixes #3623

### CODE REVIEW

**Comment by maxkatz6** [comment]
Can you add a simple unit test for this scenario?
It should be enough to check if `shape.IsMeasureValid` property returns `false` value.

**Comment by SKProCH** [comment]
@maxkatz6 done

**Comment by maxkatz6** [comment]
@SKProCH ok... I have checked a little, and noticed that change notification should have already been working with this code:
https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Media/PolylineGeometry.cs#L100-L107

But as you can see, it only supported IAvaloniaList. So, this PR should instead fix PolylineGeometry/PolygonGeometry not respecting INotifyCollectionChanged.

**Comment by SKProCH** [comment]
The problem with this geometry is under the hood implementation. We create a new Points collection inside geometry ctor:

https://github.com/AvaloniaUI/Avalonia/blob/4c949ee628ae5394e85003ff64d054b26d32c3f4/src/Avalonia.Base/Media/PolylineGeometry.cs#L46-L48

Points its just a sealed inheritor of AvaloniaList:

https://github.com/AvaloniaUI/Avalonia/blob/4c949ee628ae5394e85003ff64d054b26d32c3f4/src/Avalonia.Base/Points.cs#L6-L16

But as we can see inside AvaloniaList's ctor, it copying collection content from the collection parameter:

https://github.com/AvaloniaUI/Avalonia/blob/4c949ee628ae5394e85003ff64d054b26d32c3f4/src/Avalonia.Base/Collections/AvaloniaList.cs#L76-L79

So, at the moment where we create an Points collection inside of Geometry ctor we just copying the elements of the original collection, and we don't have reference to original collection and the changes doesn't propagated to the Points. 

It's actually has the same effect if we just call .ToList() to original collection. 

So, currently there is no actual benefits of changing OnPointsChanged implementation as you suggest.

**Comment by maxkatz6** [comment]
Try to replace
```
return new PolylineGeometry(Points, true);
```
with 
```
return new PolylineGeometry
{
    Points = Points,
    IsFilled = true
}
```

File: src/Avalonia.Controls/Shapes/Shape.cs [hunk 1]
```diff
@@ -225,12 +232,29 @@ protected static void AffectsGeometry<TShape>(params AvaloniaProperty[] properti
...
+        }
 
         /// <summary>
         /// Invalidates the geometry of this shape.
         /// </summary>
         protected void InvalidateGeometry()
         {
+            if (_definingGeometry is not null)
+            {
+                _definingGeometry.Changed += GeometryChangedHandler;
```
Comment (maxkatz6) [inline src/Avalonia.Controls/Shapes/Shape.cs:255]:
```suggestion
                _definingGeometry.Changed -= GeometryChangedHandler;
```

File: src/Avalonia.Controls/Shapes/Shape.cs [hunk 2]
```diff
@@ -75,6 +76,10 @@ public abstract class Shape : Control
                 if (_definingGeometry == null)
                 {
                     _definingGeometry = CreateDefiningGeometry();
+                    if (_definingGeometry is not null)
+                    {
+                        _definingGeometry.Changed += GeometryChangedHandler;
+                    }
```
Comment (maxkatz6) [inline src/Avalonia.Controls/Shapes/Shape.cs:82]: Probably needs to be subscribed only if attached to the visual tree.

**Comment by SKProCH** [comment]
Actually some of tests failing, i'll fix that a little later

**Review by maxkatz6 — APPROVED** [review]

**Pull request merged**